activerecord/lib/active_record/associations/association_scope.rb

```diff
@@ -158,6 +158,20 @@ def add_constraints(scope, owner, association_klass, refl, chain_head, chain_tai
         scope
       end
 
+      def merge_joins(scope, other)
+        joins_dependency, rest = other.joins_values.partition do |join|
```

Comment:
the same code we have here https://github.com/rails/rails/blob/fix/joins-in-association-conditions/activerecord/lib/active_record/relation/merger.rb#L113
I think it will be good to extract it somewhere